File: src/zeit/processing/day_summarizer.py
import logging
from datetime import datetime

from pydantic import BaseModel

from zeit.core.activity_types import ExtendedActivity
from zeit.core.llm_provider import LLMProvider
from zeit.core.prompts import DAY_SUMMARIZATION_PROMPT, DAY_SUMMARIZATION_WITH_OBJECTIVES_PROMPT
from zeit.data.db import ActivityEntry, DayObjectives
from zeit.processing.activity_summarization import ActivityGroup, build_condensed_summary

logger = logging.getLogger(__name__)


class DaySummary(BaseModel):
    summary: str
    percentages_breakdown: str
    start_time: datetime
    end_time: datetime


class DaySummarizer:
    def __init__(self, provider: LLMProvider) -> None:
        self.provider = provider

    def _format_time_range(self, start: datetime, end: datetime) -> str:
        """Format a time range like '09:15-09:45' or just '09:15' if single minute."""
        if start == end:
            return start.strftime("%H:%M")
        return f"{start.strftime('%H:%M')}-{end.strftime('%H:%M')}"

    def _format_group(self, group: ActivityGroup) -> str:
        """Format an activity group for the prompt."""
        time_range = self._format_time_range(group.start_time, group.end_time)
        reasoning = "; ".join(group.reasonings) if group.reasonings else "No description"
        activity_name = group.activity.value.replace("_", " ")
        return f'{time_range} - {activity_name} ({group.duration_minutes} min): "{reasoning}"'

    def _format_objectives_section(self, objectives: DayObjectives) -> str:
        """Format secondary objectives section for the prompt."""
        if not objectives.secondary_objectives:
            return ""
        lines = ["**Secondary Objectives:**"]
        for obj in objectives.secondary_objectives:
            lines.append(f"- {obj}")
        return "\n".join(lines)

    def summarize(
        self,
        activities: list[ActivityEntry],
        objectives: DayObjectives | None = None,
    ) -> DaySummary | None:
        non_idle = [a for a in activities if a.activity != ExtendedActivity.IDLE]

        if not non_idle:
            return None

        logger.info(f"Starting summarization with {len(non_idle)} non-idle activities")
        if objectives:
            logger.info(f"Using objectives: main='{objectives.main_objective}'")
            secondary = (
                "\n".join(objectives.secondary_objectives)
                if objectives.secondary_objectives
                else "None"
            )
            logger.info(f"Secondary objectives: {secondary}")

        # Build condensed summary with grouped activities
        condensed = build_condensed_summary(entries=activities)

        logger.info(
            f"Condensed {condensed.original_entry_count} activities "
            f"into {condensed.condensed_entry_count} groups"
        )

        # Format condensed activities for prompt
        activities_text = "\n".join(self._format_group(group) for group in condensed.groups)

        # Format percentage breakdown
        percentage_text = "\n".join(
            f"- {p.activity.value.replace('_', ' ')}: {p.percentage:.1f}%"
            for p in condensed.percentage_breakdown
        )

        # Build the final prompt based on whether objectives are provided
        if objectives:
            prompt = DAY_SUMMARIZATION_WITH_OBJECTIVES_PROMPT.format(
                main_objective=objectives.main_objective,
                secondary_objectives_section=self._format_objectives_section(objectives),
                percentage_breakdown=percentage_text,
                activities_text=activities_text,
            )
        else:
            prompt = DAY_SUMMARIZATION_PROMPT.format(
                percentage_breakdown=percentage_text,
                activities_text=activities_text,
            )
        logger.debug(f"Day summarization prompt:\n{prompt}")

        try:
            response_text = self.provider.generate(prompt, temperature=0.7)
        except Exception as e:
            logger.error(f"Failed to summarize day activities: {e}", exc_info=True)
            return None

        start_time = datetime.fromisoformat(non_idle[0].timestamp)
        end_time = datetime.fromisoformat(non_idle[-1].timestamp)

        logger.debug(f"Day summary generated for {start_time.date()}")
        return DaySummary(
            summary=response_text,
            percentages_breakdown=percentage_text,
            start_time=start_time,
            end_time=end_time,
        )

File: Sources/ZeitApp/LLM/DaySummarizer.swift
import Foundation
import os

private let logger = Logger(subsystem: "com.zeit", category: "DaySummarizer")

/// Result of summarizing a day's activities
struct DaySummary: Sendable {
    let summary: String
    let objectivesAlignment: String?
    let percentagesBreakdown: String
    let startTime: Date
    let endTime: Date
}

/// Summarizes a day's activities using an LLM
struct DaySummarizer: Sendable {
    private let provider: LLMProvider

    init(provider: LLMProvider) {
        self.provider = provider
    }

    /// Summarize a day's activities, optionally considering objectives.
    ///
    /// - Parameters:
    ///   - activities: All activity entries for the day.
    ///   - objectives: Optional day objectives.
    /// - Returns: A `DaySummary`, or `nil` if there are no non-idle activities.
    func summarize(
        activities: [ActivityEntry],
        objectives: DayObjectives? = nil,
        activityTypes: [ActivityType] = ActivityType.defaultTypes
    ) async throws -> DaySummary? {
        let nonIdle = activities.filter { $0.activity != .idle }

        guard !nonIdle.isEmpty else {
            return nil
        }

        logger.info("Starting summarization with \(nonIdle.count) non-idle activities")

        // Build condensed summary with grouped activities
        let condensed = buildCondensedSummary(from: activities, activityTypes: activityTypes)

        logger.info(
            "Condensed \(condensed.originalEntryCount) activities into \(condensed.condensedEntryCount) groups"
        )

        // Format condensed activities for prompt
        let activitiesText = condensed.groups
            .map { formatGroup($0) }
            .joined(separator: "\n")

        // Format percentage breakdown
        let percentageText = condensed.percentageBreakdown
            .map { "- \($0.activity.displayName.lowercased()): \(String(format: "%.1f", $0.percentage))%" }
            .joined(separator: "\n")

        // Build the prompt
        let objectivesTuple: (main: String, secondary: [String])?
        if let obj = objectives {
            objectivesTuple = (main: obj.mainObjective, secondary: obj.secondaryObjectives)
        } else {
            objectivesTuple = nil
        }

        let prompt = Prompts.daySummary(
            activitiesText: activitiesText,
            percentageBreakdown: percentageText,
            objectives: objectivesTuple
        )

        logger.debug("Day summarization prompt:\n\(prompt)")

        let schema = Self.summarySchema(hasObjectives: objectives != nil)

        let responseText = try await provider.generateStructured(
            prompt: prompt,
            schema: schema,
            temperature: 0.7
        )

        let parsed = try parseSummaryResponse(responseText)

        let isoFormatter = ISO8601DateFormatter()
        let startTime = isoFormatter.date(from: nonIdle[0].timestamp) ?? Date()
        let endTime = isoFormatter.date(from: nonIdle[nonIdle.count - 1].timestamp) ?? Date()

        logger.debug("Day summary generated")
        return DaySummary(
            summary: parsed.summary,
            objectivesAlignment: parsed.objectivesAlignment,
            percentagesBreakdown: percentageText,
            startTime: startTime,
            endTime: endTime
        )
    }

    // MARK: - JSON Schema

    private static func summarySchema(hasObjectives: Bool) -> [String: Any] {
        var properties: [String: Any] = [
            "summary": [
                "type": "string",
                "description": "A concise 2-3 sentence narrative summary of the day's activities"
            ]
        ]
        var required = ["summary"]

        if hasObjectives {
            properties["objectives_alignment"] = [
                "type": "string",
                "description": "1-2 sentence assessment of how well the day's activities aligned with the stated objectives"
            ]
            required.append("objectives_alignment")
        }

        return [
            "type": "object",
            "properties": properties,
            "required": required
        ]
    }

    // MARK: - Response Parsing

    private func parseSummaryResponse(_ response: String) throws -> SummaryResponse {
        let trimmed = response.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let data = trimmed.data(using: .utf8) else {
            throw DaySummarizerError.invalidResponse("Could not encode response as UTF-8")
        }
        do {
            return try JSONDecoder().decode(SummaryResponse.self, from: data)
        } catch {
            throw DaySummarizerError.invalidResponse(
                "Failed to parse summary: \(error.localizedDescription). Response was: \(trimmed.prefix(200))"
            )
        }
    }

    // MARK: - Formatting

    private func formatTimeRange(start: Date, end: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        if start == end {
            return formatter.string(from: start)
        }
        return "\(formatter.string(from: start))-\(formatter.string(from: end))"
    }

    private func formatGroup(_ group: ActivityGroup) -> String {
        let timeRange = formatTimeRange(start: group.startTime, end: group.endTime)
        let reasoning = group.reasonings.isEmpty
            ? "No description"
            : group.reasonings.joined(separator: "; ")
        let activityName = group.activity.displayName.lowercased()
        return "\(timeRange) - \(activityName) (\(group.durationMinutes) min): \"\(reasoning)\""
    }
}

// MARK: - Response Models

private struct SummaryResponse: Decodable {
    let summary: String
    let objectivesAlignment: String?

    enum CodingKeys: String, CodingKey {
        case summary
        case objectivesAlignment = "objectives_alignment"
    }
}

// MARK: - Errors

enum DaySummarizerError: LocalizedError {
    case invalidResponse(String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse(let message):
            return "Invalid summary response: \(message)"
        }
    }
}
